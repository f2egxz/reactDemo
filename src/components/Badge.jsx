//-*- mode: react -*-
//-*- coding: utf-8 -*-

/**
 * Badge.jsx
 *
 * 徽标组件
 *
 *
 * -TODO 小圆点的徽标
 * -TODO 带数字的徽标，并且可以设置封顶数值
 * -TODO 'success'、'processing'、'default'、'error'、'warning' 五种状态点
 *
 * @author fe.gxz
 */

import React          from 'react'
import { classnames } from 'util/view-extra'
import Styles         from 'style/Badge.css'


function MapStatusToStyle (dot,status){
  if(dot){
    switch (status){
      case 'Success':
        return classnames(Styles.badgeBase,Styles.badgeBaseDot,Styles.success);
      case 'Processing':
        return classnames(Styles.badgeBase,Styles.badgeBaseDot,Styles.processing);
      case 'Default':
        return classnames(Styles.badgeBase,Styles.badgeBaseDot,Styles.default);
      case 'Error':
        return classnames(Styles.badgeBase,Styles.badgeBaseDot);
      case 'Warning':
        return classnames(Styles.badgeBase,Styles.badgeBaseDot,Styles.warning);
      default:
        return new TypeError(`Unknow status type: ${status}`)
    }
  }else {
    return classnames(Styles.badgeBase,Styles.badgeNum)
  }
}



function Badge({
  count,
  overflowCount,
  dot,
  status,
  children
}){
  const num =dot?'':count>overflowCount?overflowCount+'+':count
  const Badges = (
    <div className={Styles.box}>
      {children}
      <span className={MapStatusToStyle(dot,status)}>{num}</span>
    </div>
  )
  return Badges
}

Badge.defaultProps = {
  dot : false,
  status : 'Error',
  overflowCount : 99
}

Badge.PropTypes = {
  count:          React.PropTypes.number || React.PropTypes.string,
  overflowCount:  React.PropTypes.number || React.PropTypes.string,
  dot:            React.PropTypes.bool,
  status: React.PropTypes.oneOf([
    'success',
    'processing',
    'default',
    'error',
    'warning'
  ])
}


export default Badge
