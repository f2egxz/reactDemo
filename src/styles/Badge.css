/*-*- mode: css -*-*/
/*-*- coding: utf-8 -*-*/

/**
 * Badge.css
 *
 * 徽标数
 *
 * @author fe.gxz
 */

.box {
  position: relative;
  min-width: 30px;
  min-height: 30px;
  background-color: #ddd;
  border-radius: 4px;
  display: inline-block;
}
.badgeBase{
  position: absolute;
  top:0;
  right: 0;
  transform: translate(50%,-50%);
}

.badgeNum {
  border-radius: 10px;
  background: #F04134;
  border: 1px solid transparent;
  color: #fff;
  line-height: 18px;
  text-align: center;
  padding: 0 6px;
  font-size: 12px;
  white-space: nowrap;
  box-shadow: 0 0 0 1px #fff;
}

.badgeBaseDot {
  width: 6px;
  height: 6px;
  border-radius: 100%;
  background: #F04134;
  border: 1px solid transparent;
  white-space: nowrap;
  box-shadow: 0 0 0 1px #fff;
}

.success {
  background: #00A854;
}

.processing {
  background: #108ee9;
  animation: StatusProcessing 1.2s infinite ease-in-out;
}

.default {
  background: #d9d9d9;
}

.warning {
  background: #ffbf00;
}

@keyframes StatusProcessing {
  0% ,100%{
    opacity: 1;
  }
  50% {
    opacity: 0;
  }
}
