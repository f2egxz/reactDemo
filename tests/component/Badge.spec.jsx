import React from 'react'
import Badge from 'component/Badge.jsx'
import render from 'react-test-renderer'

test('Badge',()=>{
  const component1 = render.create(
    <Badge count={100} overflowCount="10">
      <a href="##" style={{ display: 'block',backgroundColor: '#00b400', width: '60px',
        height: '60px'}}>test</a>
    </Badge>
  );
  const component2 = render.create(
    <Badge dot={true} status="Processing"></Badge>
  );
  let BadgeString = component1.toJSON();
  expect(BadgeString).toMatchSnapshot();

  BadgeString = component2.toJSON();
  expect(BadgeString).toMatchSnapshot();
})
