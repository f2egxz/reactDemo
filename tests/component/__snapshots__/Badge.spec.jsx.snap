exports[`test Badge 1`] = `
<div
  className="box">
  <a
    href="##"
    style={
      Object {
        "backgroundColor": "#00b400",
        "display": "block",
        "height": "60px",
        "width": "60px",
      }
    }>
    test
  </a>
  <span
    className="badgeBase badgeNum">
    10+
  </span>
</div>
`;

exports[`test Badge 2`] = `
<div
  className="box">
  <span
    className="badgeBase badgeBaseDot processing">
    
  </span>
</div>
`;
